p2p/src/sync/peer.rs

```diff
@@ -357,4 +428,36 @@ where
             | P2pError::InvalidStorageState(_)) => Err(e),
         }
     }
+
+    /// Sends a block list request.
+    ///
+    /// The number of headers sent equals to `P2pConfig::requested_blocks_limit`, the remaining
+    /// headers are stored in the peer context.
+    fn request_blocks(&mut self, mut headers: Vec<BlockHeader>) -> Result<()> {
+        debug_assert!(self.known_headers.is_empty());
+        if headers.len() > *self.p2p_config.max_request_blocks_count {
+            self.known_headers = headers.split_off(*self.p2p_config.max_request_blocks_count);
+        }
+
+        let block_ids: Vec<_> = headers.into_iter().map(|h| h.get_id()).collect();
+        self.message_sender.send((
+            self.id(),
+            SyncMessage::BlockListRequest(BlockListRequest::new(block_ids.clone())),
+        ))?;
+        self.requested_blocks.extend(block_ids);
```

Comment:
Maybe we should first check which blocks have already been requested?
Otherwise we might request some blocks twice and the peer will be penalized because of this check:

```
        if self.requested_blocks.take(&block.get_id()).is_none() {
            return Err(P2pError::ProtocolError(ProtocolError::UnexpectedMessage(
                "block response",
            )));
        }
```

Reply:
`self.requested_blocks` is a BTreeSet... I don't think this is possible. Unless I'm missing the point.

Reply:
I think it can be done this way:
```rust
headers.retain(|h| self.requested_blocks.contains(&h.get_id()));
```